From 83af590138bc79dfee8df8d2ec46c2bce510c5d5 Mon Sep 17 00:00:00 2001
From: Gu Zitao <guzitao@wxiat.com>
Date: Thu, 10 Nov 2022 08:44:07 +0800
Subject: [PATCH] sw64: ftrace: fix ARCH_SUPPORTS_FTRACE_OPS support
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 83af590138bc79dfee8df8d2ec46c2bce510c5d5
Modified-by-SEL: No


Sunway inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I645RV

--------------------------------

ARCH_SUPPORTS_FTRACE_OPS support was introduced with livepatch,
but it didn't pass function_trace_op to tracer in ftrace_caller,
and kernel will crash in ftrace func_profiler test.

To fix this problem, pass function_trace_op to tracer as the 3rd
argument.

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Reviewed-by: He Sheng <hesheng@wxiat.com>
Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/kernel/entry-ftrace.S | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/sw_64/kernel/entry-ftrace.S b/arch/sw_64/kernel/entry-ftrace.S
index 1b8896d1f631..6e659bff5775 100644
--- a/arch/sw_64/kernel/entry-ftrace.S
+++ b/arch/sw_64/kernel/entry-ftrace.S
@@ -130,6 +130,7 @@ ftrace_caller:
 
 	subl	$28, MCOUNT_INSN_SIZE, $16
 	bis	$26, $31, $17
+	ldl	$18, function_trace_op
 
 	ldi	$4, current_tracer
 	ldl	$27, 0($4)
-- 
2.33.0

